Bigquery output Plugin

gkatzioura opened the pull request:
### Required for all PRs:

The pull request for the Issue #8630

- [x] Signed [CLA](https://influxdata.com/community/cla/).
- [x] Associated README.md updated.
- [x] Has appropriate unit tests.

Jagularr commented on plugins/outputs/bigquery/bigquery.go:
```diff
@@ -0,0 +1,215 @@
+	return b.setUpDefaultClient()
+}
+
+func (b *BigQuery) setUpTestClient(endpoint string) error {
```
Feels like this method could be moved to the test file

Jagularr reviewed (commented):
Looks solid to me

srebhan commented on plugins/outputs/bigquery/bigquery.go:
```diff
@@ -0,0 +1,215 @@
+	"reflect"
+
+	"cloud.google.com/go/bigquery"
+	"github.com/influxdata/telegraf"
```
You might want to put this one in a separate "group" below as it is an "internal" package.

srebhan commented on plugins/outputs/bigquery/bigquery.go:
```diff
@@ -0,0 +1,215 @@
+	return b.setUpDefaultClient()
+}
+
+func (b *BigQuery) setUpTestClient(endpoint string) error {
+	noAuth := option.WithoutAuthentication()
+	endpoints := option.WithEndpoint("http://" + endpoint)
+
+	ctx := context.Background()
+
+	if c, err := bigquery.NewClient(ctx, b.Project, noAuth, endpoints); err != nil {
+		return err
+	} else {
+		b.client = c
+		return nil
+	}
+}
```
Can you please move this to the testing! You can guard the actual function by a check for `b.client` not being `nil` if you have to.

srebhan commented on plugins/outputs/bigquery/bigquery.go:
```diff
@@ -0,0 +1,215 @@
+		return fmt.Errorf("Dataset is a required field for BigQuery output")
+	}
+
+	b.setUpClient()
```
For testing you could do
```suggestion
	// To allow testing we do not override the client. The client will never be initialized in
	// normal environment. 
	if b.client == nil {
		b.setUpClient()
	}
```
and init `b.Client` with what you have in `setUpTestClient()`. This way you can move all

srebhan commented on plugins/outputs/bigquery/bigquery.go:
```diff
@@ -0,0 +1,215 @@
+
+	if c, err := bigquery.NewClient(ctx, b.Project, noAuth, endpoints); err != nil {
+		return err
+	} else {
+		b.client = c
+		return nil
+	}
```
As you return in the if you can save some indention:
```suggestion
	}
	b.client = c

	return nil
```

srebhan commented on plugins/outputs/bigquery/bigquery.go:
```diff
@@ -0,0 +1,215 @@
+func (b *BigQuery) setUpDefaultClient() error {
+	var credentialsOption option.ClientOption
+
+	ctx := context.Background()
```
Maybe you want a user-definable timeout and use it here!? Do not forget to `defer cancel()` in this case.

srebhan replied:
Please mark as resolved.

srebhan commented on plugins/outputs/bigquery/bigquery.go:
```diff
@@ -0,0 +1,215 @@
+
+	for k, v := range groupedMetrics {
+		if err := b.insertToTable(k, v); err != nil {
+			return err
```
This means you drop all remaining metrics. Skipping over this erroneous one but process the others might be better
```suggestion
			b.Log.Errorf("inserting metric %q failed: %v", <metric id/name>, err)
			continue
```

srebhan commented on plugins/outputs/bigquery/bigquery.go:
```diff
@@ -0,0 +1,215 @@
+	groupedMetrics := b.groupByMetricName(metrics)
+
+	for k, v := range groupedMetrics {
+		if err := b.insertToTable(k, v); err != nil {
```
Does it make sense to do this in parallel using a go-function?

gkatzioura replied:
I updated it, to achieve it through a Wait Group.

srebhan replied:
Please mark as resolved.

srebhan commented on plugins/outputs/bigquery/bigquery.go:
```diff
@@ -0,0 +1,215 @@
+}
+
+func (b *BigQuery) insertToTable(metricName string, metrics []bigquery.ValueSaver) error {
+	ctx := context.Background()
```
Again, you might want a user-definable timeout...

srebhan replied:
Please mark as resolved.

srebhan commented on plugins/outputs/bigquery/bigquery.go:
```diff
@@ -0,0 +1,215 @@
+	return b.client.Close()
+}
+
+// SampleConfig returns the formatted sample configuration for the plugin.
+func (s *BigQuery) SampleConfig() string {
+	return sampleConfig
+}
+
+// Description returns the human-readable function definition of the plugin.
+func (s *BigQuery) Description() string {
+	return "Configuration for Google Cloud BigQuery to send entries"
+}
```
Usually those are at the top.

srebhan commented on plugins/outputs/bigquery/bigquery.go:
```diff
@@ -0,0 +1,215 @@
+func (b *BigQuery) Close() error {
+	return b.client.Close()
+}
+
```
You are missing a `init()` function to register this output plugin.

srebhan reviewed (changes requested):
Hey this looks quite good already. My main concerns are

1. You are missing an `init()` function to register the plugin.
2. You might want a user-definable `timeout` option to avoid blocking forever in operations.
3. You should move the testing machinery out of the main plugin file.
4. Please add `cloud.google.com/go/bigquery` to `docs/LICENSE_OF_DEPENDENCIES.md`.

See also my comments in the code.

srebhan commented on plugins/outputs/bigquery/bigquery.go:
```diff
@@ -0,0 +1,233 @@
+const sampleConfig = `	
+  ## Credentials File
+  credentials_file = "/path/to/service/account/key.json"
+  ## GCP Project
```
Could you please avoid abbreviations here as some first-time-users might have difficulties to understand what it is about... :-)

srebhan commented on plugins/outputs/bigquery/bigquery.go:
```diff
@@ -0,0 +1,233 @@
+	return nil
+}
+
+func (b *BigQuery) setUpTimeout() {
+	if b.Timeout.Duration < 1*time.Second {
+		b.Timeout.Duration = defaultTimeout
+	}
+}
```
Could you please also move this to `Connect()` to have a nice consistent structure. It will add to the beauty of this code.

srebhan commented on plugins/outputs/bigquery/bigquery.go:
```diff
@@ -0,0 +1,233 @@
+  ## The namespace for the metric descriptor
+  dataset = "telegraf"
+
+  ## Timeout for BigQuery operations.  Minimum timeout is "1s".
```
Is there a reason you enforce a timeout of one second? If it is about default-value handling, the easy solution is to set the default in `init()` when initializing the structure. This way the value is only overwritten when set by the user.

gkatzioura replied:
Hi @srebhan no specific reason. Was just checking the zookeeper connector that did the same. Did the default as you proposed.

srebhan commented on plugins/outputs/bigquery/bigquery.go:
```diff
@@ -0,0 +1,233 @@
+
+	for k, v := range groupedMetrics {
+		wg.Add(1)
+		go b.insertToTable(k, v, &wg)
```
You might use an anonymous function here to put the `defer wg.Done()` here rather than the function body for everybody to see it's done correctly. But that is a matter of taste and I leave it to you.

gkatzioura replied:
Moved to anonymous function

srebhan commented on plugins/outputs/bigquery/bigquery_test.go:
```diff
@@ -0,0 +1,96 @@
+
+var testingHost string
+
+func TestMain(t *testing.M) {
+	srv := localBigQueryServer(t)
+
+	testingHost = strings.ReplaceAll(srv.URL, "http://", "")
+
+	defer srv.Close()
+
+	os.Exit(t.Run())
+}
```
I don't get the intention behind this test... If you only want to test if you can setup a local test-server, I think you don't need to do it.

gkatzioura replied:
I run the tests through TestMain. I use TestMain in order to spin up the server that produces a fake BigQuery Server. Then through T.Run() the tests `TestWrite` and  `TestConnect` will run using the fake BigQuery Server.

srebhan replied:
This is an unusual pattern for telegraf. :-) Is there any reason not to spin up a server in every test (that's what everybody else in telegraf does)?

gkatzioura replied:
Changed to spinning up per test.

srebhan commented on plugins/outputs/bigquery/bigquery_test.go:
```diff
@@ -0,0 +1,96 @@
+	err := b.Write(mockMetrics)
+	require.NoError(t, err)
+}
+
```
Could you add a test where you actually compare what is sent to the server with what you expect? Do you have an example of the on-wire format?

srebhan reviewed (commented):
Very nice code! Thank you! Only few comments in the code. One nice-to-have thing is a test where you actually compare what is sent to the server with your expectations. But I'm not sure if this is doable!?

srebhan commented:
Hey @gkatzioura you are writing the code faster than I can review! ;-) Good job! :-)

srebhan commented on plugins/outputs/bigquery/bigquery.go:
```diff
@@ -110,7 +110,10 @@ func (b *BigQuery) Write(metrics []telegraf.Metric) error {
-		go b.insertToTable(k, v, &wg)
+		go func(k string, v []bigquery.ValueSaver) {
+			b.insertToTable(k, v)
+			wg.Done()
```
Make it `defer wg.Done()` and move it to the top of the anonymous function so we can be sure it will be done in any case.

srebhan commented on plugins/outputs/bigquery/bigquery.go:
```diff
@@ -0,0 +1,229 @@
+
+	for k, v := range groupedMetrics {
+		wg.Add(1)
+		go func(k string, v []bigquery.ValueSaver) {
+			defer wg.Done()
+			b.insertToTable(k, v)
+		}(k, v)
```
This should also work
```suggestion
		go func() {
			defer wg.Done()
			b.insertToTable(k, v)
		}()
```

srebhan commented:
@gkatzioura can you please comment on why you enforce a minimum timeout of 1s!?

gkatzioura commented:
> @gkatzioura can you please comment on why you enforce a minimum timeout of 1s!?

Hi @srebhan will do, bit busy now due to work.

srebhan commented:
@gkatzioura ok don't worry, just liked the live coding/review in the morning... ;-)

srebhan commented on plugins/outputs/bigquery/bigquery.go:
```diff
@@ -0,0 +1,227 @@
+	timeStampFieldName = "timestamp"
+)
+
+var defaultTimeout = internal.Duration{Duration: time.Duration(time.Second * 5)}
```
Can you please do:
```suggestion
var defaultTimeout = internal.Duration{Duration: 5 * time.Second}
```
as there is a big PR fixing those superfluous expressions (#8835) and Pawel will not be happy if we add more. ;-)

srebhan commented on plugins/outputs/bigquery/bigquery_test.go:
```diff
@@ -21,6 +23,7 @@ const (
 )
 
 var testingHost string
+var testDuration = internal.Duration{Duration: time.Duration(time.Second * 5)}
```
Please remove `time.Duration`.

srebhan commented:
Almost. Please remove the `time.Duration` as the timeout is already a duration. Furthermore, you can remove`cloud.google.com/go/bigquery` from the list-of-dependencies document as `cloud.google.com/go/*` is covered already.

Last thing is the test-structure. Can you please start the server in each test individually to match "the usual pattern"!?

srebhan commented on plugins/outputs/bigquery/README.md:
```diff
@@ -0,0 +1,46 @@
+
+### Restrictions
+
+Avoid hyphens on BigQuery tables, underlying sdk cannot handle straming inserts to Table with hyphens.
```
```suggestion
Avoid hyphens on BigQuery tables, underlying SDK cannot handle streaming inserts to Table with hyphens.
```?

srebhan commented on plugins/outputs/bigquery/README.md:
```diff
@@ -0,0 +1,46 @@
+* string
+* boolean
+
+All field naming restrictions that apply to BigQuery sould apply to the measurements to be imported.
```
```suggestion
All field naming restrictions that apply to BigQuery should apply to the measurements to be imported.
```

srebhan commented on plugins/outputs/bigquery/README.md:
```diff
@@ -0,0 +1,46 @@
+
+Tables on BigQuery should be created before hand and they are not created during persistence
+
+Pay attention to the columnd `timestamp` since it is reserved upfront and cannot change. 
```
```suggestion
Pay attention to the column `timestamp` since it is reserved upfront and cannot change. 
```?

srebhan commented on plugins/outputs/bigquery/README.md:
```diff
@@ -0,0 +1,46 @@
+
+All field naming restrictions that apply to BigQuery sould apply to the measurements to be imported.
+
+Tables on BigQuery should be created before hand and they are not created during persistence
```
```suggestion
Tables on BigQuery should be created beforehand and they are not created during persistence
```

srebhan commented on plugins/outputs/bigquery/bigquery.go:
```diff
@@ -0,0 +1,227 @@
+	"github.com/influxdata/telegraf/plugins/outputs"
+)
+
+const (
+	defaultOffSetKey   = "offset-key.json"
+	timeStampFieldName = "timestamp"
+)
```
I think this isn't used.
```suggestion
const timeStampFieldName = "timestamp"
```

srebhan commented on plugins/outputs/bigquery/bigquery.go:
```diff
@@ -0,0 +1,227 @@
+
+const sampleConfig = `	
+  ## Credentials File
+  credentials_file = "/path/to/service/account/key.json"
+  ## Google Cloud Platform Project
```
```suggestion
  credentials_file = "/path/to/service/account/key.json"

  ## Google Cloud Platform Project
```

srebhan commented on plugins/outputs/bigquery/bigquery.go:
```diff
@@ -0,0 +1,227 @@
+	CredentialsFile string `toml:"credentials_file"`
+	Project         string `toml:"project"`
+	Dataset         string `toml:"dataset"`
+
```
```suggestion
```

srebhan commented on plugins/outputs/bigquery/bigquery.go:
```diff
@@ -0,0 +1,227 @@
+	Dataset         string `toml:"dataset"`
+
+	Timeout internal.Duration `toml:"timeout"`
+
+	Log telegraf.Logger
```
```suggestion
	Log telegraf.Logger `toml: "-"`
```

srebhan reviewed (commented):
I went through everything a (almost) last time and found some more typos and formatting things. Only cosmetics. After fixing those I think we are good to go.

gkatzioura commented:
Hi @srebhan I added a map in order to do a lookup for cases of metrics with hyphens. Due to the SDK and the api having issues inserting to tables with hyphens this connector would have been limited to only non-hyphen metrics. 
Instead by having this map the user can define an association between a metric with hyphens and the corresponding table.

srebhan commented:
I'm a bit ambivalent here. On the one hand-side I think letting the user specify a replacement (e.g. underscore, space or nothing, i.e. empty) would be good enough in my view, on the other hand this is not much code. What makes me tend to the "user specifies a replacement" (with a sensible default) solution is the fact that this way you can _ensure_ that you can process the metric.
In the table case, the user might simply forget, mistype or do some other error that leaves you with a hyphen-metric without a mapping in the table. What do you do in such cases?

Any thoughts?

gkatzioura commented:
Hi @srebhan. 
My concern was users dropping the plugin due to the naming limitation, since many measurement will have hyphens.If the users specify a map (for the hyphens cases, there is no need to do it for everything, the connector should also work out of the box) they are covered on this issue but as you mentioned they are not covered in case they forget that they have a hyphen metric.
It seems ok in case of a hyphen metric where the user has not specified a table mapping, to fallback on removing the hyphens (and print a warning). 
Thus co-existence of table mapping exists and also a fallback.
However the user still needs to create the BigQuery tables upfront. So somehow he still needs to be aware on how to name the tables and there also might be a rename collision.

srebhan commented:
@gkatzioura I can follow your argumentation, but a mixture of concepts (replacement of hyphen vs. replacement of metric name) is always the worst thing to do... ;-)

> If the users specify a map (for the hyphens cases, there is no need to do it for everything, the connector should also work out of the box) they are covered on this issue but as you mentioned they are not covered in case they forget that they have a hyphen metric.

Right. That's my concern. In large installations with many plugins and maybe multiple users modifying the config you will hit the  "forgotten mapping" case with certainty! That's why I would prefer to specify a replacement character (e.g. underscore) for the hyphens. If the user wants to really rename the metrics he/she can still use the [rename processor](https://github.com/influxdata/telegraf/tree/master/plugins/processors/rename) for doing so. This would keep the config options for your plugins short (it's just so beautifully brief!) and makes sure we cover _all_ cases by default.

gkatzioura commented:
Hi @srebhan I think the [rename processor](https://github.com/influxdata/telegraf/tree/master/plugins/processors/rename) is amazing and there no need for what I did with the map (was not fully aware :/ on it). Thus I shall remove the map mechanism.

On your comment you proposed to swap hyphens with underscores. Are you ok for me to proceed?

srebhan commented:
@gkatzioura yeah that's my proposal. However, I would like to see a config option with the replacement, having underscore as default and removing hyphens if empty. Do you think this makes sense?

For logging on replace, please track which metric you already warned on to avoid spamming the log-file. You can do this easily by using a `map[string]bool` that you set if you warned and before warning check `if ! alreadyWarned[metric]`...

gkatzioura commented:
Hi @srebhan will proceed as suggested.

gkatzioura commented:
Hi @srebhan did proceed on the changes and updated the Readme.
* Added a default hyphen replacement
* Added configuration on the character replacing
* Print warning only once

srebhan commented on plugins/outputs/bigquery/bigquery.go:
```diff
@@ -0,0 +1,248 @@
+
+var defaultTimeout = internal.Duration{Duration: 5 * time.Second}
+
+const sampleConfig = `	
+  ## Credentials File
+  credentials_file = "/path/to/service/account/key.json"
+
+  ## Google Cloud Platform Project
+  project = "my-gcp-project"
+
+  ## The namespace for the metric descriptor
+  dataset = "telegraf"
+
+  ## Timeout for BigQuery operations.
+  # timeout = "5s"
+`
```
This misses the `replace_hyphen_to` option. Please also sync with `README.md`.

srebhan commented on plugins/outputs/bigquery/bigquery.go:
```diff
@@ -0,0 +1,248 @@
+}
+
+func (b *BigQuery) metricToTable(metricName string) string {
+	if strings.Contains(metricName, "-") {
+
+		dhm := strings.ReplaceAll(metricName, "-", b.ReplaceHyphenTo)
+
+		if warned, _ := b.warnedOnHyphens[metricName]; !warned {
+			b.Log.Warn("Metric %q contains hyphens please consider using the rename processor plugin, falling back to %q", dhm)
+			b.warnedOnHyphens[metricName] = true
+		}
+
+		return dhm
+	}
+
+	return metricName
```
Please invert the logic to save indention levels:
```suggestion
	if ! strings.Contains(metricName, "-") {
		return metricName
	}

	dhm := strings.ReplaceAll(metricName, "-", b.ReplaceHyphenTo)
	if warned, _ := b.warnedOnHyphens[metricName]; !warned {
		b.Log.Warn("Metric %q contains hyphens please consider using the rename processor plugin, falling back to %q", dhm)
		b.warnedOnHyphens[metricName] = true
	}

	return dhm
```

srebhan commented on plugins/outputs/bigquery/bigquery.go:
```diff
@@ -0,0 +1,248 @@
+	outputs.Add("bigquery", func() telegraf.Output {
+		return &BigQuery{
+			Timeout:         defaultTimeout,
+			warnedOnHyphens: make(map[string]bool),
```
Please do this in `Init()`.

gkatzioura replied:
I moved this above. Did you mean doing the map initialization on the Connect()?

srebhan replied:
I mean moving `b.warnOnHyphens = make(map[string]bool)` to `Connect()` instead of doing it here. Sorry for the confusion, `Init()` is the `Connect()` in input plugins... :-)

srebhan reviewed (commented):
Very good, please check the minor comments and update `sampleConfig` and `README.md` to the new option(s).

srebhan reviewed (approved):
This code looks beautiful IMO. Thank you very much for your effort!

srebhan commented:
Oh just one more request @gkatzioura... Please add this plugin to the [top-level README file's output list](https://github.com/influxdata/telegraf/blob/master/README.md#output-plugins).

srebhan reviewed (approved):
Perfect.

srebhan commented:
@gkatzioura please resolve the merge conflicts!

srebhan reviewed (approved):
It's off again, please take a look from time to time so this one stays merge-able. I'll try to find out when this might be merged.

ssoroka commented:
I'm good to merge this after you've had a look at my last two comments.

gkatzioura commented on plugins/outputs/bigquery/bigquery.go:
```diff
@@ -0,0 +1,251 @@
+	table := b.client.DatasetInProject(b.Project, b.Dataset).Table(tableName)
+	inserter := table.Inserter()
+
+	if err := inserter.Put(ctx, metrics); err != nil {
+		b.Log.Errorf("inserting metric %q failed: %v", metricName, err)
+	}
```
Yes this is a no retry case. In case of a failure due to the availability of BigQuery the assumption is that the connector is misconfigured (BigQuery Table does not exist, wrong fields). Let me know what would work better.

ssoroka replied:
I'm fine with it if you are. I do wonder if maybe the Goroutines aren't really necessary. The retry logic would be easier without them, as on 5xx status code you'd return an error, and on 4xx you'd log it and return nil. If they're all going over the same client connection, I think they're being serialized anyway.

gkatzioura replied:
Yes, initially the pr was not using Goroutines, but I was encouraged to parallelize the inserts. I will raise a PR removing them, thus more control and will return on error.

ssoroka commented:
I'm going to merge and you can adjust the goroutines if you want in a new PR. [See my last reply](https://github.com/influxdata/telegraf/pull/8634#discussion_r593324308)
